File: Sdks/Java/sender/src/main/java/metis/client/sdk/sender/BaseHttpSender.java
package metis.client.sdk.sender;

import metis.client.sdk.SingleSender;
import metis.client.sdk.client.GathererClient;
import metis.client.sdk.entity.LogEntity;
import metis.client.sdk.utility.CsvSerializer;
import metis.client.sdk.utility.UriPath;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.net.URLEncoder;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentLinkedQueue;

/**
 * Created by Administrator on 14-9-5.
 */
public class BaseHttpSender implements SingleSender {

    private static final Logger logger = LoggerFactory.getLogger(BaseHttpSender.class);

    protected int maxLogEntry = 5000;
    protected int interval = 3;

    protected final int batchSize = 200;
    protected final int sendBatchSize = 500;

    protected CsvSerializer serializer = new CsvSerializer();
    protected ConcurrentLinkedQueue<LogEntity> logList = new ConcurrentLinkedQueue<LogEntity>();
    protected GathererClient httpClient = new GathererClient(2000);
    protected String gathererPath = "";

    @Override
    public void prepare(Map<String, Object> config) {
        if(config.get("GathererPath") == null)
            throw new IllegalArgumentException("必须存在GathererPath的配置");
        this.gathererPath = String.valueOf(config.get("GathererPath"));
        //设置HttpClient的参数
        Object gathererHost = null;
        if(config.containsKey("GathererHost")) {
            httpClient.setHost(String.valueOf(config.get("GathererHost")));
        }
        httpClient.setHeader("charset", "utf-8");
        //设置其他的变量
        if(config.containsKey("MaxQueueSize")) {
            this.maxLogEntry = Integer.parseInt(config.get("MaxQueueSize").toString());
        }
        if(config.containsKey("SendInterval")) {
            this.interval = Integer.parseInt(config.get("SendInterval").toString());
        }
    }

    @Override
    public void doAppend(LogEntity entry) {
        if(logList.size() < this.maxLogEntry) {
            logList.add(entry);
        }
    }

    @Override
    public void clear() {

    }

    protected void postData(List<LogEntity> logs) {
        try {
            if(logs.size() == 0) {
                return;
            }
            LogEntity firstLog = logs.get(0);
            Map<String, String> data = new HashMap<String, String>();
            data.put("TerminalCode", String.valueOf(firstLog.getTerminalCode()));
            String sendData = serializer.serializeMutil(firstLog.getClass(), logs);
            data.put("MultiData", URLEncoder.encode(sendData, "UTF-8"));
            //获取需要添加的路径
            String postUrl = getRequestUrl(firstLog.getLogType().toLowerCase());
            httpClient.setHeader("AccessToken", firstLog.getAccessToken());
            //向服务器POST数据
            httpClient.httpPost(postUrl, data);
            //响应正确计数
            for(LogEntity log : logs) {
                log.ack();
            }
        } catch(Exception ex) {
            if(logger.isErrorEnabled()) {
                logger.error("HTTPSENDER", ex);
            }
            //响应失败计数
            for(LogEntity log : logs) {
                log.fail();
            }
        }
    }

    private String getRequestUrl(String... args) {
        List<String> frames = new ArrayList<String>();
        frames.add(gathererPath);
        for(String s : args) {
            frames.add(s);
        }
        return UriPath.combine((String[])frames.toArray());
    }
}

File: Sdks/DotNet/Metis.ClientSdk.Sender/Senders/BaseHttpSender.cs
using Metis.ClientSdk.Entities;
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Collections.Specialized;
using System.Linq;
using System.Text;
using System.Web;

namespace Metis.ClientSdk.Sender
{
    internal abstract class BaseHttpSender : ISingleSender
    {
        protected int maxLogEntry = 5000;
        protected int interval = 3;

        protected readonly int batchSize = 200;
        protected readonly int sendBatchSize = 500;

        protected readonly CsvSerializer serializer = new CsvSerializer();
        protected GaeaHttpClient httpClient = new GaeaHttpClient(2000);
        protected string gathererPath = String.Empty;

        public abstract void DoAppend(LogEntity entry);
        public virtual void Prepare(IDictionary<string, object> config)
        {
            if (!config.Keys.Contains("GathererPath"))
                throw new ArgumentNullException("必须存在GathererPath的配置");

            this.gathererPath = config["GathererPath"].ToString();
            //设置HttpClient的参数
            object gathererHost = null;
            if (config.TryGetValue("GathererHost", out gathererHost))
                httpClient.Host = gathererHost.ToString();
            httpClient.Headers.Add("charset", "utf-8");
            //设置其他变量
            object objMaxEntry = null;
            if (config.TryGetValue("MaxQueueSize", out objMaxEntry))
                Int32.TryParse(objMaxEntry.ToString(), out maxLogEntry);
            object objInterval = null;
            if (config.TryGetValue("SendInterval", out objInterval))
                Int32.TryParse(objInterval.ToString(), out interval);
        }
        public virtual void Clear()
        {
        }
        /// <summary>
        /// 发送数据
        /// </summary>
        /// <param name="logs"></param>
        protected void PostData(List<LogEntity> logs)
        {
            try
            {
                if (logs.Count == 0)
                    return;

                NameValueCollection data = new NameValueCollection();
                var firstLog = logs.FirstOrDefault();
                //添加需要发送的数据
                data.Add("TerminalCode", firstLog.TerminalCode.ToString());
                var sendData = serializer.SerializeMutil(firstLog.GetType(), logs);
                data.Add("MultiData", HttpUtility.UrlEncode(sendData));
                //获取需要添加的路径
                var postUrl = GetRequestUrl(firstLog.LogType.ToLower());
                if (httpClient.Headers.AllKeys.Contains("AccessToken"))
                {
                    httpClient.Headers["AccessToken"] = firstLog.AccessToken;
                }
                else
                {
                    httpClient.Headers.Add("AccessToken", firstLog.AccessToken);
                }
                //向服务器POST数据
                httpClient.HttpPost(postUrl, data);
                //响应正确
                logs.ForEach(o => o.Ack());
            }
            catch
            {
                //响应失败
                logs.ForEach(o => o.Fail());
            }
        }
        private string GetRequestUrl(params string[] args)
        {
            var frames = new List<string>();
            frames.Add(gathererPath);
            frames.AddRange(args);
            return UriPath.Combine(frames.ToArray());
        }
    }
}
